### DISCUSSION TITLE
Distributed Notification Names

### DISCUSSION THREAD

**Discussion by FelixKratz** [FelixKratz/SketchyBar#151] [General]
Lets collectively create a list of NSDistributedNotificationCenter
notifications we can hook onto via the event system. Please comment any you
come across and I will update the list in this post:

| <NSDistributedNotificationCenterName>             | Description                                                          |
| --                                                | --                                                                   |
| AppleInterfaceThemeChangedNotification            | The system theme changes between light and dark mode                 |
| AppleInterfaceMenuBarHidingChangedNotification    | The auto-hide menu bar setting is changed                            |
| AppleSelectedInputSourcesChangedNotification      | The keyboard input source changes                                    |
| com.apple.bluetooth.status                        | A bluetooth device is connected/disconnected                         |
| com.apple.screenIsLocked                          | The screen is locked                                                 |
| com.apple.screenIsUnlocked                        | The screen is unlocked                                               |
| com.apple.HIToolbox.beginMenuTrackingNotification | The mouse begins tracking a menu, e.g. if a menu bar item is clicked |
| com.apple.HIToolbox.endMenuTrackingNotification   | The mouse ends tracking a menu, e.g. if a menu bar item is clicked   |
| com.apple.sharedfilelist.change                   | A new file is included in the shared file list                       |
| com.apple.DownloadFileFinished                    | A download has finished                                              |
| com.apple.unmountassistant.process.start          | A hard drive is about to get mounted                                 |
| com.apple.LaunchServices.applicationRegistered    | A new application is registered for the auto start                   |
| _NSDoNotDisturbEnabledNotification                | The system enters do not disturb mode                                |
| _NSDoNotDisturbDisabledNotification               | The system exits do not disturb mode                                 |
| com.spotify.client.PlaybackStateChanged           | Spotify changes its current track                                    |
| com.apple.Music.playerInfo                        | The music app changes its current track                              |

**Comment by rhyses-pieces** [comment]
I found this [handy link](https://developer.apple.com/documentation/foundation/nsnotificationname) that lists _a lot_ of NSNotificationNames that may or may not work with the eventing system.

[NSCalendarDayChangedNotification](https://developer.apple.com/documentation/foundation/nscalendardaychangednotification) looks useful for simple calendar scripting. Will do some testing!

**Comment by bphenriques** [comment]
Bluetooth:
- Device changed: `com.apple.bluetooth.status`
- Bluetooth on: `IOBluetoothHostControllerPoweredOnNotification`
- Bluetooth off: `IOBluetoothHostControllerPoweredOffNotification`

Example:
```
bluetooth=(
  icon.font="$FONT_FACE:Regular:19.0"
  label.padding_left=0
  label.padding_right=0
  script="$PLUGIN_DIR/bluetooth.sh"
)

...

  --add item bluetooth right \
    --add event bluetooth_status "com.apple.bluetooth.status" \
    --add event bluetooth_on "IOBluetoothHostControllerPoweredOnNotification" \
    --add event bluetooth_off "IOBluetoothHostControllerPoweredOffNotification" \
    --set bluetooth "${bluetooth[@]}" \
    --subscribe bluetooth system_woke bluetooth_status bluetooth_off bluetooth_on \
```

Noticed a small delay when disconnecting a headset, but it does work in few seconds.

Edit: for some reason, it stopped working
Edit2: Now works. Had a problem with missing `jq` in the `PATH` leading to bluetooth script not working as expected.

**Reply by Fau818** [reply]
Thank you, it is very helpful!

**Reply by diontr00** [reply]
bluetooth on and off events does not seem to work anymore ,are  u experienced the same problem ?

**Comment by Pe8er** [comment]
```
_NSDoNotDisturbEnabledNotification
_NSDoNotDisturbDisabledNotification
```

These don't seem to work anymore in Sonoma.

**Comment by marsjane** [comment]
```#!/bin/sh

BLUETOOTH_EVENT="com.apple.bluetooth.status"
sketchybar --add event bluetooth_update $BLUETOOTH_EVENT

sketchybar --add item  airpods.right right                            \
           --set       airpods.right script="$PLUGIN_DIR/airpods.sh"  \
                               update_freq=10                         \
                               padding_left=1                         \
                               padding_right=2                        \
                               background.border_width=0              \
                               background.height=24                   \
                               icon=􀲋                                 \
                               icon.color=$GREEN                      \
                               label.color=$WHITE                     \
           --subscribe airpods.right bluetooth_update volume_change   \
           --add item  airpods.case right                             \
           --set       airpods.case script="$PLUGIN_DIR/airpods.sh"   \
                                  update_freq=10                      \
                                  padding_left=1                      \
                                  padding_right=1                     \
                                  background.border_width=0           \
                                  background.height=24                \
                                  icon=􀹫                              \
                                  icon.color=$GREEN                   \
                                  label.color=$WHITE                  \
           --subscribe airpods.case bluetooth_update volume_change    \
           --add item  airpods.left right                             \
           --set       airpods.left script="$PLUGIN_DIR/airpods.sh"   \
                                  update_freq=10                      \
                                  padding_left=8                      \
                                  padding_right=1                     \
                                  background.border_width=0           \
                                  background.height=24                \
                                  icon=􀲌                              \
                                  icon.color=$GREEN                   \
                                  label.color=$WHITE                  \
           --subscribe airpods.left bluetooth_update volume_change    \


sketchybar --add bracket status brew github.bell wifi volume_icon airpods.right airpods.case airpods.left \
           --set status "${status_bracket[@]}" \
           --subscribe status bluetooth_update volume_change
```
Got one issue that when I connect my airpods, these several items not triggered anymore. it was normal if I totally restart the sketchybar.
Also, here I set the update_freq is 10 seconds, why it doesn't rerun each 10 seconds? I guess if it rerun 10sseconds, it should appear even not triggered by the event

**Reply by marsjane** [reply]
Ok I fixed this by adding the `updates=on` for these items, I am using FelixKratz's setting and I just noticed the default updates value is when_shown. I should read the doc more carefully 😆

**Comment by RaiSaugat** [comment]
Will I be able to get the battery percent of the connected headset? My headset is not from apple. It's anker soundcore q45. I want to know if we can show the battery percent or not?

**Comment by theevilbit** [comment]
You can get many names just by checking the logs.

```
user@mac % log show --predicate 'eventMessage contains[c] "register name"' --last 1h | awk '{ print $12 }' | sort | uniq -c | sort
   1 
   1 CPInstalledProvisionProfilesDidChangeNotification
   1 HelpBookRegistrationDidChange
   1 SafeBrowsingEnabledStateDidChangeNotification
   1 UADomainPointerControlSettingsDidChangeNotification
   1 UniversalAccessDomainCloseViewSettingsDidChangeNotification
   1 UniversalAccessDomainKeyboardSettingsDidChangeNotification
   1 UniversalAccessDomainMouseSettingsDidChangeNotification
   1 UniversalAccessDomainSpeechSettingsDidChangeNotification
   1 com.apple.CFNetwork.CookiesChanged.2fd9b6fcf2aefaa49880952bcbdb8831c43697c6
   1 com.apple.CFNetwork.CookiesChanged.99be6d30b7ea4722d30f502e79d3b84e30499b89
   1 com.apple.ForceSuppressedDidChangeNotification
   1 com.apple.LaunchServices.pluginswillberemoved
   1 com.apple.UAContrastDidChange
   1 com.apple.accessibility.hasUserCustomFontsDidChange
   1 kDictationIMNotificationUseOnlyOfflineDictationChanged
   2 AutoRecordingEnableChanged
   2 CPInstalledProfilesDidChangeNotification
   2 MMPrefsChanged
   2 NSFavoriteStylesChanged
   2 NSPersistentUIBitmapEncryptionKeyDidChange
   2 NSWindowForceInvalidateShadow
   2 PARBagChangeNotification
   2 com.apple.DownloadFileChanged
   2 com.apple.LaunchServices.SystemConfigurationChange
   2 com.apple.accessibility.AXVisualSupportAgent.hoverTextSettingsDidChange
   2 com.apple.screensaver.didlaunch
   2 com.apple.scriptmenu
   2 com.apple.sharpointmanagment.change
   2 com.apple.trackpad.OrientationModeDidChange
   2 com.apple.unmountassistant.process.end
   2 com.apple.unmountassistant.process.start
   2 kFSFileManagerVolumePropertyChanged
   2 sidebarMiscVisibilityChanged
   3 ABDefaultAccountDidChangeDistributedNotification
   3 ABDistributedAccountConfigurationDidChangeNotification
   3 AppleSideBarDefaultIconSizeChanged
   3 DVTDiagnosticsGatherNotification
   3 DVTDiagnosticsRunPhasesNotification
   3 DVTDiagnosticsRunPhasesTimeoutNotification
   3 DVTPlatformAddedDownloadedSDKNotification
   3 DictationIMNotificationStartedListening
   3 NSAutomaticCapitalizationEnabledChanged
   3 NSAutomaticDashSubstitutionEnabledChanged
   3 NSAutomaticInlinePredictionEnabledChanged
   3 NSAutomaticPeriodSubstitutionEnabledChanged
   3 NSAutomaticQuoteSubstitutionEnabledChanged
   3 NSAutomaticSpellingCorrectionEnabledChanged
   3 NSAutomaticTextCompletionCollapsedChanged
   3 NSAutomaticTextCompletionEnabledChanged
   3 NSSpellServerLanguagesChanged
   3 NSSpellServerQuotesChanged
   3 NSSpellServerReplacementsChanged
   3 NSUserReplacementItemsEnabledChanged
   3 PHXDistributedSourceAddedNotification
   3 PHXDistributedSourceChangedNotification
   3 PHXDistributedSourceRemovedNotification
   3 PHXDistributedSourcesInvalidatedNotification
   3 com.apple.UserPictureDidChangeNotification
   3 com.apple.shortcuts.WFCoreDataDatabaseContextDidSaveNotification
   4 NSServicesChangedNotification
   4 com.apple.AOSKit.AccountChanged
   4 com.apple.MCX._managementStatusChangedForDomains
   5 QLSharedPreviewPanelWillOpenNotification
   5 com.apple.LaunchServices.pluginsregistered
   5 com.apple.LaunchServices.pluginsunregistered
   6 AppleAquaScrollBarVariantChanged
   6 AppleGestureScrollSettingChanged
   6 AppleShowScrollBarsSettingChanged
   7 ACDAccountStoreDidChangeNotification
   7 NSSecondaryDisplayVisChanged
   7 sidebarSharedSectionEnabledChanged
   8 com.apple.HIToolbox.cancelMenuTrackingNotification
   9 kFinderPreferencesChangedNotification
  10 com.apple.quicklook.GlobalChange
  12 ABDistributedDatabaseChangedNotification
  12 ABDistributedDatabaseChangedNotificationPriv
  12 ABWillRestoreFromMetadataPriv
  16 com.apple.inputmethodkit.abnormaldeath
  20 com.apple.LaunchServices.ShowAllExtensionsChangedToFalse
  20 com.apple.LaunchServices.ShowAllExtensionsChangedToTrue
  58 DictationIMNotificationDidExitDictationMode
  58 IronwoodNotificationHotKeyChanged
  62 com.apple.nsquiet_safe_quit_give_reason
  65 com.apple.userkeyequivalentsdidchange
  75 AppleDatePreferencesChangedNotification
  75 AppleEnabledInputSourcesChangedNotification
  75 AppleInstalledInputSourcesChangedNotification
  75 AppleKeyboardMenuPreferencesChangedNotification
  75 AppleKeyboardPreferencesChangedNotification
  75 AppleKeyboardPreferencesSelectedItemsChangedNotification
  75 AppleRomanSwitchSettingChangedNotification
  75 AppleSelectedInputMethodPasteboardInvalidNotification
  75 AppleSelectedInputSourcesChangedNotification
  75 AppleTextBehaviorPreferencesChangedNotification
  75 AppleTextInputMenuExtraDidLoadNotification
  75 AppleTimePreferencesChangedNotification
  75 UADomainKeyboardAccessSettingsDidChangeNotification
  75 UniversalAccessDomainAssistiveControlSettingsDidChangeNotification
  75 com.apple.AppleSpacesSwitchOnActivate
  75 com.apple.Carbon.TISNotifyEnabledHandwritingLanguagesChanged
  75 com.apple.Carbon.TISNotifyEnabledKeyboardInputSourcesChanged
  75 com.apple.Carbon.TISNotifyIronwoodAutoEnableChanged
  75 com.apple.Carbon.TISNotifySelectedHandwritingLanguageChanged
  75 com.apple.universalaccess.VoiceOverSettingsDidChange
  76 AppleAquaColorVariantChanged
  76 AppleColorPreferencesChangedNotification
  76 AppleEnableMouseSwipeNavigateWithScrollsDidChangeNotification
  76 AppleEnableSwipeNavigateWithScrollsDidChangeNotification
  76 AppleInterfaceMenuBarHidingChangedNotification
  76 ApplePrivateInterfaceThemeChangedNotification
  76 NSEventResetDefaults
  76 com.apple.AppleMultitouchTrackpadDomainDidChangeNotification
  76 com.apple.sharedfilelist.change
  76 com.apple.sharedfilelist.launch
  76 com.apple.xctest.FakeForceTouchDevice
  77 AppleAppSleepPreferencesChangedNotification
  77 AppleNumberPreferencesChangedNotification
  77 AppleReduceDesktopTintingChanged
  77 com.apple.accessibility.api
  82 AppleLanguagePreferencesChangedNotification
  82 com.apple.KeyboardUIModeDidChange
  88 com.apple.Carbon.TISNotifySelectedKeyboardInputSourceChanged
 106 AXInterfaceClassicInvertColorStatusDidChange
 106 AXInterfaceDifferentiateWithoutColorStatusDidChange
 106 AXInterfaceReduceMotionStatusDidChange
 106 AXInterfaceReduceTransparencyStatusDidChange
 106 AXInterfaceShowToolbarButtonShapesStatusDidChangeNotification
 106 AXInterfaceShowWindowTitlebarIconsStatusDidChangeNotification
 107 com.apple.universalaccess.screenGrayscaleDidChange
 107 com.apple.universalaccess.screenPolarityDidChange
 132 AppleAquaAntiAliasingChanged
 132 UAPMenuFontSizeDidChangeNotification
 150 AppleKeyboardPreferencesFnKeySettingChangedNotification
 150 com.apple.Carbon.TISNotifyHandwritingInputSourceChanged
 153 __NSFakeMemoryPressureNotification
 181 AXInterfaceIncreaseContrastStatusDidChange
 225 AppleNoRedisplayAppearancePreferenceChanged
```

log format is:

```
2024-01-31 09:36:06.395565+0100 0x6d58f3   Default     0x0                  582    0    distnoted: [com.apple.distnoted:diagnostic] register name: AXInterfaceShowWindowTitlebarIconsStatusDidChangeNotification object: kCFNotificationAnyObject token: a00000009 pid: 30826
```

**Comment by Fau818** [comment]
Is there a Distributed Notification about changes in output devices?

I would like to change my volume icon to `AirPods` when the output device is my AirPods.

**Reply by Nikolaidp24** [reply]
I was trying to do the same thing and didn't find anything on the notification's side, so I tried writing a function in bash to switch audio and send event trigger that way, but ultimately just resorted back to using alias...

**Reply by diontr00** [reply]
As far as i experience is that  Bluetooth Event are  not working anymore on sonoma , before i can hook into IOBluetoothHostControllerPoweredOnNotificastion and com.apple.controlcenter with SwitchAudioSource to make the feature

**Comment by nima9** [comment]
if anyone knows if there is an event that fires on volume/mute change, please let me know!

**Reply by FelixKratz** [reply]
There is an integrated event in sketchybar for that: `volume_change`: https://felixkratz.github.io/SketchyBar/config/events

**Comment by diontr00** [comment]
These bluettoth events are not work anymore  (Sonoma 14.2):
IOBluetoothHostControllerPoweredOnNotification
IOBluetoothHostControllerPoweredOffNotification
When bluetooth turn on and  off those  events are not triggered , anyone experience the  same problem ?

**Comment by rogeruiz** [comment]
I've noticed that the `com.apple.Music.playerInfo` event stopped firing / being recognized by SketchyBar on macOS Sonoma. I use it as a backup for `media_changed` events not firing / capturing the `$INFO`. 😕

**Comment by MrJarnould** [comment]
Has anyone found a notification that detects changes between different Focus modes? While `NSDoNotDisturbEnabledNotification` effectively indicates when *any* Focus mode is activated and `_NSDoNotDisturbDisabledNotification` signals when *any* Focus mode is deactivated, I am unable to locate a notification specifically for detecting transitions between individual Focus modes.

**Reply by rogeruiz** [reply]
I haven't tried this but you can determine the focus mode using JSX. Here's a gist that can help you get started. 

https://gist.github.com/drewkerr/0f2b61ce34e2b9e3ce0ec6a92ab05c18

**Comment by alexcamlo** [comment]
Has someone found a notification for input mute or not? Trying to have the mic_status plugin in SbarLua.

**Comment by Masstronaut** [comment]
Has anyone found a notification for when a diplay/monitor is connected or disconnected?
